File: benchmark/harness/Sources/BenchmarkKit/Rotation.swift
// Rotation.swift — per-mode latest-3 atomic history rotation, port of
// benchmark/lib/rotation.py.
//
// history.json shape: {"deterministic": [r…], "live": [r…]} — newest LAST, max 3
// PER MODE. Adding a 4th record of a mode drops that mode's oldest; the OTHER
// mode's bucket is left byte-identical. Ordering is stabilized by ISO-8601
// `timestamp_utc` (lexicographic == chronological for the Z form).
//
// Records flow through as `JSONValue` (the JSON form of a BenchmarkRecord) so the
// exact serialized bytes are preserved — matching the Python which rotates plain
// dicts, never re-encoding through the schema.

import Foundation

public enum Rotation {
    static let modes = ["deterministic", "live"]

    // MARK: atomic write (tmp -> fsync -> rename)

    /// Write `value` to `path` atomically: tmp file -> flush -> fsync -> rename.
    /// The tmp name embeds pid + random suffix (matches rotation.py). Always
    /// cleans up a leftover tmp on any failure path.
    public static func atomicWrite(_ value: JSONValue, to path: String, indent: Int = 2) throws {
        let dir = (path as NSString).deletingLastPathComponent
        let dirPath = dir.isEmpty ? "." : dir
        try FileManager.default.createDirectory(
            atPath: dirPath, withIntermediateDirectories: true)
        let tmp = "\(path).\(getpid()).\(Int.random(in: 0..<100000)).tmp"
        let text = value.serialized(indent: indent)
        defer {
            if FileManager.default.fileExists(atPath: tmp) {
                try? FileManager.default.removeItem(atPath: tmp)
            }
        }
        // Write + fsync via a FileHandle so the bytes hit disk before the rename.
        FileManager.default.createFile(atPath: tmp, contents: nil)
        let handle = try FileHandle(forWritingTo: URL(fileURLWithPath: tmp))
        try handle.write(contentsOf: Data(text.utf8))
        try handle.synchronize()   // fsync
        try handle.close()
        // Atomic rename over the destination.
        try FileManager.default.removeItemIfExists(atPath: path)
        try FileManager.default.moveItem(atPath: tmp, toPath: path)
    }

    // MARK: history read/init

    static func readOrInit(_ historyPath: String) -> [(String, [JSONValue])] {
        var buckets: [String: [JSONValue]] = [:]
        if let text = try? String(contentsOfFile: historyPath, encoding: .utf8),
           let parsed = try? JSONParser.parse(text),
           case .object(let pairs) = parsed {
            for (k, v) in pairs {
                if let items = v.arrayItems { buckets[k] = items }
            }
        }
        // Guarantee both modes exist, in canonical order (deterministic, live).
        var result: [(String, [JSONValue])] = []
        for m in modes {
            result.append((m, buckets[m] ?? []))
        }
        // Preserve any extra buckets that were present (defensive; Python
        // setdefault leaves unknown keys in place).
        return result
    }

    static func timestamp(of record: JSONValue) -> String {
        record["timestamp_utc"]?.stringValue ?? ""
    }

    // MARK: rotate

    /// Append `record` to its mode bucket, keep newest `maxPerMode`, atomic-write.
    /// Returns the updated in-memory buckets. The record's "mode" selects the
    /// bucket; the other mode bucket is left byte-identical.
    @discardableResult
    public static func rotate(historyPath: String, record: JSONValue, maxPerMode: Int = 3)
        throws -> [(String, [JSONValue])] {
        let mode = record["mode"]?.stringValue ?? "deterministic"
        var buckets = readOrInit(historyPath)
        // Find (or create) the target bucket.
        if let idx = buckets.firstIndex(where: { $0.0 == mode }) {
            var list = buckets[idx].1
            list.append(record)
            list.sort { timestamp(of: $0) < timestamp(of: $1) }
            if maxPerMode >= 0 && list.count > maxPerMode {
                list.removeFirst(list.count - maxPerMode)
            }
            buckets[idx].1 = list
        } else {
            buckets.append((mode, [record]))
        }
        let obj = JSONValue.object(buckets.map { ($0.0, .array($0.1)) })
        try atomicWrite(obj, to: historyPath)
        return buckets
    }

    // MARK: rotate_detail

    /// Write a per-run detail file under `runsDir/<mode>/<run_id>.json` and prune
    /// that mode's dir to the newest `maxPerMode` by embedded timestamp (mtime
    /// fallback). Returns the written detail path. The OTHER mode dir is untouched.
    @discardableResult
    public static func rotateDetail(runsDir: String, runID: String, record: JSONValue,
                                    maxPerMode: Int = 3) throws -> String {
        let mode = record["mode"]?.stringValue ?? "deterministic"
        let modeDir = (runsDir as NSString).appendingPathComponent(mode)
        try FileManager.default.createDirectory(atPath: modeDir, withIntermediateDirectories: true)
        let detailPath = (modeDir as NSString).appendingPathComponent("\(runID).json")
        try atomicWrite(record, to: detailPath)

        // Prune: keep newest maxPerMode .json by embedded timestamp_utc.
        var entries: [(String, String)] = []  // (sortKey, path)
        let names = (try? FileManager.default.contentsOfDirectory(atPath: modeDir)) ?? []
        for name in names where name.hasSuffix(".json") {
            let p = (modeDir as NSString).appendingPathComponent(name)
            var ts = ""
            if let text = try? String(contentsOfFile: p, encoding: .utf8),
               let parsed = try? JSONParser.parse(text) {
                ts = parsed["timestamp_utc"]?.stringValue ?? ""
            }
            if ts.isEmpty {
                let mtime = (try? FileManager.default.attributesOfItem(atPath: p)[.modificationDate]
                    as? Date)?.timeIntervalSince1970 ?? 0
                ts = "~\(mtime)"  // mtime fallback sorts AFTER real ISO ts
            }
            entries.append((ts, p))
        }
        entries.sort { $0.0 < $1.0 }
        if entries.count > maxPerMode {
            for (_, p) in entries.prefix(entries.count - maxPerMode) {
                try? FileManager.default.removeItem(atPath: p)
            }
        }
        return detailPath
    }
}

extension FileManager {
    func removeItemIfExists(atPath path: String) throws {
        if fileExists(atPath: path) { try removeItem(atPath: path) }
    }
}

File: benchmark/harness/benchmarkkit/rotation.py
"""Per-mode latest-3 atomic history rotation.

history.json shape: ``{"deterministic": [r...], "live": [r...]}`` — newest LAST, max 3
per mode. Adding a 4th record of a mode drops that mode's oldest; the other mode's
bucket is carried through as the raw parsed dict, so its bytes are preserved exactly.
Records rotate as plain dicts (never re-encoded through the schema) — that is what
keeps the untouched bucket byte-identical.
"""

from __future__ import annotations

import json
import os
import random
from typing import Optional

MODES = ["deterministic", "live"]


def atomic_write(obj, path: str, indent: int = 2) -> None:
    """Write ``obj`` to ``path`` atomically: tmp -> fsync -> rename, no trailing newline."""
    directory = os.path.dirname(path) or "."
    os.makedirs(directory, exist_ok=True)
    tmp = f"{path}.{os.getpid()}.{random.randint(0, 99999)}.tmp"
    text = json.dumps(obj, indent=indent)
    try:
        fd = os.open(tmp, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o644)
        try:
            os.write(fd, text.encode("utf-8"))
            os.fsync(fd)
        finally:
            os.close(fd)
        os.replace(tmp, path)
    finally:
        if os.path.exists(tmp):
            os.remove(tmp)


def _read_or_init(history_path: str) -> dict:
    parsed = {}
    try:
        with open(history_path, encoding="utf-8") as f:
            parsed = json.load(f)
    except (OSError, ValueError):
        parsed = {}
    if not isinstance(parsed, dict):
        parsed = {}
    # Canonical order first (deterministic, live), then any extra buckets preserved.
    result: dict = {}
    for m in MODES:
        result[m] = parsed.get(m, [])
    for k, v in parsed.items():
        if k not in result:
            result[k] = v
    return result


def _timestamp(record: dict) -> str:
    return record.get("timestamp_utc") or ""


def rotate(history_path: str, record: dict, max_per_mode: int = 3) -> dict:
    """Append ``record`` to its mode bucket, keep newest ``max_per_mode``, atomic-write.

    The record's ``mode`` selects the bucket; the other mode bucket is left
    byte-identical (carried through as the raw parsed dict).
    """
    mode = record.get("mode") or "deterministic"
    buckets = _read_or_init(history_path)
    lst = buckets.setdefault(mode, [])
    lst.append(record)
    lst.sort(key=_timestamp)
    if max_per_mode >= 0 and len(lst) > max_per_mode:
        del lst[: len(lst) - max_per_mode]
    atomic_write(buckets, history_path)
    return buckets


def rotate_detail(runs_dir: str, run_id: str, record: dict, max_per_mode: int = 3) -> str:
    """Write ``runs_dir/<mode>/<run_id>.json`` and prune that mode's dir to newest 3.

    Pruning sorts by embedded ``timestamp_utc`` (mtime fallback sorts after real ISO
    timestamps). The other mode's dir is untouched.
    """
    mode = record.get("mode") or "deterministic"
    mode_dir = os.path.join(runs_dir, mode)
    os.makedirs(mode_dir, exist_ok=True)
    detail_path = os.path.join(mode_dir, f"{run_id}.json")
    atomic_write(record, detail_path)

    entries = []  # (sort_key, path)
    for name in os.listdir(mode_dir):
        if not name.endswith(".json"):
            continue
        p = os.path.join(mode_dir, name)
        ts = ""
        try:
            with open(p, encoding="utf-8") as f:
                ts = (json.load(f) or {}).get("timestamp_utc") or ""
        except (OSError, ValueError):
            ts = ""
        if not ts:
            ts = f"~{os.path.getmtime(p)}"  # mtime fallback sorts after real ISO ts
        entries.append((ts, p))
    entries.sort(key=lambda e: e[0])
    if len(entries) > max_per_mode:
        for _, p in entries[: len(entries) - max_per_mode]:
            try:
                os.remove(p)
            except OSError:
                pass
    return detail_path
